File: 课程代码/程序设计课程实践/2022程序设计实践-作业3/T2/T2.py
# 为了尽可能贴近C语言的风格，这里很多写法都比较憨
# 大家可以由C入门python,快速get用法
mp = [[] for i in range(64)]# 创建二维数组
vis = [[0 for i in range(64)] for i in range(64)] ## 创建二维数组，并都置0
n = m = ans = 0
dx = [1,-1,0,0]# 方向列表
dy = [0,0,-1,1]

def dfs(x,y,t):
    if vis[x][y]:# 如果访问过就返回
        return
    vis[x][y]=1
    for i in range(4):
        nx = x+dx[i]
        ny = y+dy[i]
        if nx>=0 and nx<n and ny>=0 and ny<m and mp[x][y]==mp[nx][ny] and vis[nx][ny]==0:
            dfs(nx,ny,t+1)

    return


if __name__ == "__main__":
    n,m = map(lambda x:int(x),input("").split())
    for i in range(n):
        mp[i] = list(map(lambda x:int(x),input("").split()))
    for i in range(n):
        for j in range(m):
            if vis[i][j]==0:
                dfs(i,j,0)
                ans+=1
    print(ans,end='')

File: 课程代码/程序设计课程实践/2022程序设计实践-作业3/T2/T2.c
#include <stdio.h>
//非常裸的一道DFS
int mp[64][64];//存放矩阵
int n,m;//矩阵大小
int vis[64][64];//记录已经访问过的位置
int ans = 0;//最终答案
int dx[5] = {0,1,-1,0,0};//方向数组
int dy[5] = {0,0,0,-1,1};
void dfs(int x,int y){
    if(vis[x][y])return;//走过就不再走了
    vis[x][y] = 1;
    for (int i = 1; i <= 4; i++){
        int nx = x+dx[i];
        int ny = y+dy[i];
        if(mp[x][y]==mp[nx][ny]&&!vis[nx][ny]&&nx>=1&&nx<=n&&ny>=1&&ny<=m){
            dfs(nx,ny);
        }
    }
    return;
}
int main(void){
    scanf("%d%d",&n,&m);
    for(int i = 1; i <= n; i++){//输入mp
        for(int j = 1;j <= m; j++){
            scanf("%d",&mp[i][j]);
        }
    }
    for(int i = 1; i <= n; i++){//遍历mp
        for(int j = 1;j <= m; j++){
            if(!vis[i][j]){//如果该位置没被访问过，那么就dfs
                dfs(i,j);
                ans++;
            }
        }
    }
    printf("%d",ans);
    return 0;
}